### PR TITLE
New keep importer

### PR DESCRIPTION
**Pull request by andrewconner**
This PR introduces a new model, `RawKeep`. A RawKeep stores everything needed to make a `Bookmark` (including context, userId, etc), and is used exclusively for imports right now.

For non-imports, nothing changes. URIs are normalized and `Bookmark`s are created immediately with no persistence and no queue. This is to keep hover / site / phone keeps as fast as possible (no need for single threaded, leader only) since they're almost always one at a time. We still use `RawBookmarkRepresentation` because `RawKeep`s are persisted, which is meaningless when the keep goes immediately from JSON to Bookmark.

For imports, we parse the JSON in `RawKeepFactory` (analogous to `RawBookmarkFactory`) in `ExtBookmarkController` and then immediately call `bookmarkInterner.persistRawKeeps`. This method loads the `RawKeep`s into the database as fast as possible, then notifies the `RawKeepImporterPlugin` that there are keeps waiting.

`RawKeepImporterPlugin` (using an actor) loads batches of `RawKeep`s, recovers context, and calls `BookmarkInterner`. Failures are marked, default tags are applied, and progress is tracked (for the front-end). Fortunately, now we can easily track how long imports are taking.

Some things I would like to do soon, but not urgent to merge:
- The API for `BookmarkInterner` has been left largely untouched, and is still a little messy, especially when dealing with privacy. I'd like to clean that up (turn privacy into an Option, with defaults chosen at the controller level).
- Keeping and importing should probably use separate APIs (instead of everything going eventually through `internRawBookmark`) to make their purpose and effects clear.
- Imports should probably be tracked in a separate table, so we can highly reliably keep track of progress and status for the batch as a whole. For the user, it would allow people to see their imports listed out.

### CODE REVIEW

File: kifi-backend/modules/common/app/com/keepit/common/store/InMemoryObjectStore.scala [hunk 1]
```diff
@@ -17,7 +17,7 @@ trait InMemoryObjectStore[A, B]  extends ObjectStore[A, B] with Logging {
   def += (kv: (A, B)) = {
     localStore += kv
     val v = if (kv._2 != null) kv._2.toString else ""
-    log.info(s"[+=] (${kv._1} -> ${v.take(50)}) localStore.size=${localStore.size}")
+    log.debug(s"[+=] (${kv._1} -> ${v.take(50)}) localStore.size=${localStore.size}")
```
Comment (andrewconner) [inline kifi-backend/modules/common/app/com/keepit/common/store/InMemoryObjectStore.scala:20]: @ymatsuda, this okay to switch to `log.debug`? It adds quite a bit of activity to the log when it's on `log.info`.

File: kifi-backend/modules/common/app/com/keepit/heimdal/HeimdalServiceClient.scala [hunk 2]
```diff
@@ -84,20 +84,22 @@ class HeimdalClientActor @Inject() (
   val serviceCluster = serviceDiscovery.serviceCluster(serviceType)
   private var events: Vector[HeimdalEvent] = Vector()
   private var closing = false
+  private var flushIsPending = false
 
   def receive = {
     case event: HeimdalEvent =>
-      log.info(s"Event added to queue: $event")
+      log.debug(s"Event added to queue: $event")
```
Comment (andrewconner) [inline kifi-backend/modules/common/app/com/keepit/heimdal/HeimdalServiceClient.scala:91]: @stkem @leogrim Okay to switch this to debug? It gets quite noisy in dev mode as info.

File: kifi-backend/modules/common/app/com/keepit/common/store/InMemoryObjectStore.scala (hunk 1 above)
Comment (ymatsuda) [inline kifi-backend/modules/common/app/com/keepit/common/store/InMemoryObjectStore.scala:20]: fine with me.

File: kifi-backend/modules/common/app/com/keepit/heimdal/HeimdalServiceClient.scala (hunk 2 above)
Comment (stkem) [inline kifi-backend/modules/common/app/com/keepit/heimdal/HeimdalServiceClient.scala:91]: Fine with me.

**Comment by andrewconner** [comment]
Of note: Previously, `KeepToCollection` (the Bookmark x Collection repo) would call `collectionRepo.collectionChanged` for every new tag so that the collection indexer could pick up the change. However, this had a couple bad side effects (namely, sending thousands of messages to Eliza for every import), so now, **you must explicitly call `collectionRepo.collectionChanged` when you change a collection**. This is the same pattern used for the `uriGraph`, but is new for collections.

**Comment by ymatsuda** [comment]
I am confused. It sounds very odd that the collectionChanged needs to call Eliza for very save. Does Eliza depends on the messages to maintain its data consistency? If so, what happens if the message is missed? At the moment the message is sent, data is not committed yet. How can Eliza use the message? 

urigraph uses the sequence numbers on bookmarks to maintain consistency. Sending message is just a hint that tells urigraph to look for new data to index. Even when this message is missed, urigraph won't loose any data. It will catch up by a periodical polling later or another message.

**Comment by andrewconner** [comment]
The extension keeps a copy of the users collections/tags, and gets updates when they change. This behavior hasn't changed (it just doesn't happen for every single save, and only happens manually).

If a message is missed, it's typically not a big deal. The extension UI will have an incorrect name, or be missing (or have an extra) tag in the list. You're right that this should probably be delayed until its committed. I'll need to think about the best way to do that, because if I removed sending to Eliza, I can see people easily forgetting to update the extension. Thoughts?

**Comment by ymatsuda** [comment]
Oh. It maintains data in the extension! I didn't think about it. Interesting.
How about using zookeeper and the sequence number?

File: kifi-backend/modules/shoebox/app/com/keepit/integrity/UriIntegrityPlugin.scala [hunk 3]
```diff
@@ -65,19 +66,31 @@ class UriIntegrityActor @Inject()(
...
+          }
 
           if (diff.contains(ktc.collectionId)) {
             val inactiveKtc = keepToCollectionRepo.getOpt(newBm.id.get, ktc.collectionId)
-            if (inactiveKtc.isDefined) inactiveKtc foreach { inactiveKtc =>
-              keepToCollectionRepo.save(ktc.copy(state = KeepToCollectionStates.INACTIVE))
-              keepToCollectionRepo.save(inactiveKtc.copy(state = KeepToCollectionStates.ACTIVE))
+            if (inactiveKtc.isDefined) {
+              inactiveKtc.foreach { inactiveKtc =>
+                collections = collections + inactiveKtc.collectionId
```
Comment (yingjieMiao) [inline kifi-backend/modules/shoebox/app/com/keepit/integrity/UriIntegrityPlugin.scala:80]: sorry for the late comment, but here do we want to add ktc.collectionId to collections too?

**Pull request merged**